Bluetooth: Mesh: use secure storage in ble mesh

alxelax opened the pull request:
PR:
 - adds dependency of the mbedtls psa usage on secure storage
 - removes PSA ITS emulator and enables usage of
   the secure storage in ble mesh bsim tests
 - enables secure storage in all ble mesh and related samples

alxelax committed:
Bluetooth: Mesh: use secure storage in ble mesh

Commit:
 - adds dependency of the mbedtls psa usage on secure storage
 - removes PSA ITS emulator and enables usage of
   the secure storage in ble mesh bsim tests
 - enables secure storage in all ble mesh and related samples

Signed-off-by: Aleksandr Khromykh <aleksandr.khromykh@nordicsemi.no>

tomi-font commented on subsys/secure_storage/Kconfig.its_store:
```diff
@@ -12,7 +12,7 @@ config SECURE_STORAGE_ITS_STORE_IMPLEMENTATION_SETTINGS
 		&& $(dt_node_has_compat,$(dt_node_parent,$(DT_STORAGE_PARTITION)),fixed-partitions)
 	imply FLASH_MAP
 	imply NVS
```
I'm wondering if it would make sense to remove this line as we are changing the paradigm from selecting to depending on...

tomi-font commented on subsys/secure_storage/Kconfig:
```diff
@@ -65,7 +65,7 @@ endif
 menuconfig SECURE_STORAGE_ITS_STORE_MODULE
 	bool "ITS store module"
-	imply FLASH # for FLASH_HAS_DRIVER_ENABLED
+	depends on FLASH # for FLASH_HAS_DRIVER_ENABLED
```
This was originally added to be able to check `FLASH_HAS_DRIVER_ENABLED` in `Kconfig.its_store`.
This line can now be removed completely as it's expected that `FLASH` will be enabled elsewhere. `FLASH_HAS_DRIVER_ENABLED` will still be checked in `Kconfig.its_store` (and `FLASH` is a prerequisite of it).

tomi-font commented on subsys/secure_storage/Kconfig.its_store:
```diff
@@ -12,7 +12,7 @@ config SECURE_STORAGE_ITS_STORE_IMPLEMENTATION_SETTINGS
 	imply FLASH_MAP
 	imply NVS
-	select SETTINGS
+	depends on SETTINGS
```
This will require changes to enable this (and `FLASH`) in the existing tests/samples under `tests/subsys/secure_storage/` and `samples/psa/`.

tomi-font replied:
I'd say rather add this as part of the already existing `depends on`, or split it into parts (e.g. putting `FLASH_HAS_DRIVER_ENABLED` separately), for consistency.

tomi-font replied:
The Kconfig options still need to be enabled in the samples, I think.

tomi-font reviewed (changes requested):
Might also be worth a migration guide entry to indicate that it's now necessary to select the affected Kconfig options for the secure storage subsystem.

Balaklaka commented on samples/psa/its/overlay-secure_storage.conf:
```diff
@@ -7,4 +7,8 @@ CONFIG_MBEDTLS_PSA_CRYPTO_C=y
 
+CONFIG_FLASH=y
+CONFIG_FLASH_MAP=y
+CONFIG_NVS=y
```
Could we use ZMS instead of NVS? Other places too.

alxelax replied:
I guess it should be done over platform overlay where ZMS is suitable.

alxelax replied:
@tomi-font, opinion?

tomi-font replied:
> Could we use ZMS instead of NVS? Other places too.

Soon: https://github.com/zephyrproject-rtos/zephyr/pull/81235

tomi-font commented on doc/releases/migration-guide-4.1.rst:
```diff
@@ -260,6 +260,20 @@ Other Subsystems
+==============
+
+* The :kconfig:option:`CONFIG_SECURE_STORAGE_ITS_STORE_MODULE` symbol no longer implies the
+  :kconfig:option:`CONFIG_FLASH` Kconfig option. Platforms using Secure storage must explicitly enable
+  :kconfig:option:`CONFIG_FLASH` Kconfig option.
```
```suggestion
  :kconfig:option:`CONFIG_FLASH` Kconfig option. Platforms using Secure storage must explicitly enable it.
```

tomi-font commented on doc/releases/migration-guide-4.1.rst:
```diff
@@ -260,6 +260,20 @@ Other Subsystems
+
+* The :kconfig:option:`CONFIG_SECURE_STORAGE_ITS_STORE_IMPLEMENTATION_SETTINGS` symbol no longer implies the
+  :kconfig:option:`CONFIG_FLASH_MAP` and :kconfig:option:`CONFIG_NVS` Kconfig options.
+  It either no longer selects :kconfig:option:`CONFIG_SETTINGS` Kconfig option.
```
```suggestion
  It no longer selects the :kconfig:option:`CONFIG_SETTINGS` Kconfig option either.
```

tomi-font commented on doc/releases/migration-guide-4.1.rst:
```diff
@@ -260,6 +260,20 @@ Other Subsystems
+  :kconfig:option:`CONFIG_FLASH_MAP` and :kconfig:option:`CONFIG_NVS` Kconfig options.
+  It either no longer selects :kconfig:option:`CONFIG_SETTINGS` Kconfig option.
+  Platforms using Secure storage must explicitly enable :kconfig:option:`CONFIG_FLASH_MAP`,
+  :kconfig:option:`CONFIG_SETTINGS` and either the :kconfig:option:`CONFIG_NVS` or
+  :kconfig:option:`CONFIG_ZMS` Kconfig options.
```
Let's make it more generic:
```suggestion
  :kconfig:option:`CONFIG_SETTINGS` and the desired storage system (e.g. :kconfig:option:`CONFIG_ZMS`).
```

tomi-font commented on doc/releases/migration-guide-4.1.rst:
```diff
@@ -260,6 +260,20 @@ Other Subsystems
+* The :kconfig:option:`CONFIG_SECURE_STORAGE_ITS_STORE_IMPLEMENTATION_SETTINGS` symbol no longer implies the
+  :kconfig:option:`CONFIG_FLASH_MAP` and :kconfig:option:`CONFIG_NVS` Kconfig options.
+  It either no longer selects :kconfig:option:`CONFIG_SETTINGS` Kconfig option.
+  Platforms using Secure storage must explicitly enable :kconfig:option:`CONFIG_FLASH_MAP`,
```
Using Secure storage with the settings-based ITS store module.

alxelax replied:
didn't get this. How the final text should look like?

tomi-font replied:
Actually I think it can be made as simple as the following as we previously refer to `CONFIG_SECURE_STORAGE_ITS_STORE_IMPLEMENTATION_SETTINGS`:
```suggestion
  Platforms using it must now explicitly enable :kconfig:option:`CONFIG_FLASH_MAP`,
```

tomi-font commented on doc/releases/release-notes-4.1.rst:
```diff
@@ -314,6 +314,15 @@ Libraries / Subsystems
 
 * Storage
 
+* Secure storage
+
+  * Removed the implicit enabling of the :kconfig:option:`CONFIG_FLASH` Kconfig option
+    by the :kconfig:option:`CONFIG_SECURE_STORAGE_ITS_STORE_MODULE` symbol.
+
+  * Removed the implicit enabling of the :kconfig:option:`CONFIG_FLASH_MAP`,
+    :kconfig:option:`CONFIG_NVS` and :kconfig:option:`CONFIG_SETTINGS` Kconfig options
+    by the :kconfig:option:`CONFIG_SECURE_STORAGE_ITS_STORE_IMPLEMENTATION_SETTINGS` symbol.
+
```
IMO this whole thing can be removed. It's only relevant to the migration guide.

sjanc commented:
#AutoPTS run zephyr nrf52 MESH/NODE/KR/BV-01-C DFUM/SR/FD/BV-01-C

codecoup-tester commented:
Scheduled PR https://github.com/zephyrproject-rtos/zephyr/pull/82319#issuecomment-2523281887, board: nrf52, estimated start time: 14:43:31, test case count: 2, estimated duration: 0:04:49<details><summary>Test cases to be run</summary>DFUM/SR/FD/BV-01-C<br>MESH/NODE/KR/BV-01-C</details>

codecoup-tester commented:
AutoPTS Bot results:
<details><summary>Failed tests (1)</summary>MESH    MESH/NODE/KR/BV-01-C            FAIL</details>
<details><summary>Successful tests (1)</summary>DFUM    DFUM/SR/FD/BV-01-C              PASS</details>

tomi-font commented on tests/subsys/secure_storage/psa/crypto/testcase.yaml:
```diff
@@ -2,8 +2,16 @@ common:
 tests:
-  secure_storage.psa.crypto.secure_storage:
-    filter: CONFIG_SECURE_STORAGE and not CONFIG_SECURE_STORAGE_ITS_STORE_IMPLEMENTATION_NONE
+  secure_storage.psa.crypto.secure_storage.settings:
+    filter: CONFIG_SECURE_STORAGE and CONFIG_SECURE_STORAGE_ITS_STORE_IMPLEMENTATION_SETTINGS
+            and CONFIG_FLASH_HAS_DRIVER_ENABLED
+    extra_args: EXTRA_CONF_FILE=overlay-secure_storage.conf;overlay-settings.conf
+    integration_platforms:
+      - native_sim
+      - nrf54l15dk/nrf54l15/cpuapp
+  secure_storage.psa.crypto.secure_storage.custom:
+    filter: CONFIG_SECURE_STORAGE and CONFIG_SECURE_STORAGE_ITS_STORE_IMPLEMENTATION_CUSTOM
+            and CONFIG_FLASH_HAS_DRIVER_ENABLED
```
This is a copy/paste fail right?

alxelax replied:
I guess twister filter doesn't work based on this option. I see tests and samples are built for platforms even without flash.

tomi-font commented on tests/subsys/secure_storage/psa/crypto/overlay-secure_storage.conf:
```diff
@@ -8,6 +8,10 @@ CONFIG_TIMER_RANDOM_GENERATOR=y
 CONFIG_COMMON_LIBC_MALLOC_ARENA_SIZE=2048
 CONFIG_MBEDTLS_PSA_CRYPTO_C=y
 
+CONFIG_FLASH=y
+CONFIG_FLASH_MAP=y
+CONFIG_NVS=y
+CONFIG_SETTINGS=y
```
This shouldn't be here? Only in `overlay-settings.conf`.

alxelax replied:
I guess `overlay-settings.conf` has been added after these changes. I'll move it there.

alxelax commented on tests/subsys/secure_storage/psa/its/testcase.yaml:
```diff
@@ -8,28 +8,32 @@ common:
 tests:
   secure_storage.psa.its.secure_storage:
-    filter: CONFIG_SECURE_STORAGE and not CONFIG_SECURE_STORAGE_ITS_STORE_IMPLEMENTATION_NONE
+    filter: CONFIG_SECURE_STORAGE and CONFIG_SECURE_STORAGE_ITS_STORE_IMPLEMENTATION_SETTINGS
+            and CONFIG_FLASH_HAS_DRIVER_ENABLED
```
because tests are built for platforms without flash, those `settings` subsystem is irrelevant.

alxelax commented:
Hi @tomi-font, seems yml files those are used to check secure storage samples or tests use twister `filter` feature that doesn't work properly.
Dependencies those were initially added (and later changed to even more strict dependencies) do not prevent building tests and samples for platforms without FLASH.
You can take a look, almost all twister CI steps cannot pass.

Since I'm not an expert in secure storage, I'd like to get any ideas how to solve this issue.
My suggestion is to use list of allowed platforms instead of filters.

tomi-font commented:
@alxelax It rather looks like the problems are caused by the changes in your PR. Failures in this PR pass locally on `main`. Also for instance a recent [fixup PR](https://github.com/zephyrproject-rtos/zephyr/pull/83454) had green CI.
Right now I don't have the bandwidth to look into fixing this myself, but I suggest you the following:
- Rebase your PR on the latest `main`.
- Don't, for instance, change the `not CONFIG_SECURE_STORAGE_ITS_STORE_IMPLEMENTATION_NONE` part of the filter by `CONFIG_FLASH_HAS_DRIVER_ENABLED` as you currently do in the samples' YAML files. It will for sure provoke failures. `CONFIG_SECURE_STORAGE_ITS_STORE_IMPLEMENTATION` has more complex filtering than just `FLASH_HAS_DRIVER_ENABLED` because I found out that just this Kconfig option wasn't enough (thus the check for DT properties).
- `secure_storage.psa.its.secure_storage.custom.store` doesn't require flash as the custom store implementation is RAM-based. Same applies to `secure_storage.psa.its.secure_storage.custom.both`.
- I'm pretty sure you don't want to use `FLASH_HAS_DRIVER_ENABLED` at all in the filters. You want to use the Kconfig option that corresponds to the ITS store module implementation that is expected to be used (ZMS or settings or just "not none"). The filtering is handled in those Kconfig options.

alxelax commented:
@tomi-font, I removed my changes from filters, but you can see they still do not work. Twister runs Secure storage tests on platforms without Flash (it was the initial reason why I added direct dependency on Flash driver presence). Those obviously are not built.
It seems to me like filters do not work properly.
Any ideas?

alxelax commented:
@tomi-font, I removed NVS and FLASH dependency for sample tests. Seems filters do not work. Now it passes for all platforms those sneak over filters. Really do not want to fix secure storage test system.

HaavardRei commented:
The commit titles says "use psa crypto storage in mesh bsim", but you change files in modules/, samples/, subsys/.../Kconfig and tests/. Maybe this can be split into three? (Modules, samples/Kconfig and bsim, or something like this)

alxelax commented:
> The commit title says "use psa crypto storage in mesh bsim", but you change files in modules/, samples/, subsys/.../Kconfig and tests/. Maybe this can be split into three? (Modules, samples/Kconfig and bsim, or something like this)

Almost all changes are related. It is not possible to remove its emulator and start using secure storage directly without changes in `modules/`, `subsys/` and `tests/`. It is possible to tract changes in `samples/` as more or less independent. However, I do not see big advantages to split it as separate commit since they do not have sense without changes in `subsys/.../Kconfig`. 

I will add more description details in commit message for better understanding.

PavelVPV commented on subsys/bluetooth/mesh/Kconfig:
```diff
@@ -1452,6 +1452,10 @@ config BT_MESH_SOL_ADV_XMIT
 
+config BT_MESH_SECURE_STORAGE
+	bool
+	depends on SECURE_STORAGE
```
I think this will not work as expected and will unconditionally force enabling SECURE_STORAGE: https://docs.zephyrproject.org/latest/build/kconfig/tips.html#select-pitfalls

alxelax replied:
Hm... but Mesh doesn't select `SECURE_STORAGE`. I do not think these pitfalls are relevant to this PR.

PavelVPV replied:
Mesh selects `BT_MESH_SECURE_STORAGE` which in my understanding will turn `SECURE_STORAGE` to `y` regardless of dependencies in secure storage.

tomi-font replied:
No it won't. It would if this line were `select SECURE_STORAGE`.

alxelax replied:
Nop. When Mesh selects `BT_MESH_SECURE_STORAGE` it just adds dependency on `SECURE_STORAGE`. Application still responsible for enabling `SECURE_STORAGE`.

alxelax replied:
This not public construction is required because KConfig syntax doesn't support smth like this `depends on SECURE_STORAGE if BT_SETTINGS`

tomi-font reviewed (approved):
Looks good! I just found this point not very clear in your commit message:
```
- adds dependency of the mbedtls psa usage on secure storage
``` 
I think I get what you mean after having reviewed the changes, but it could be made clearer IMO. Also, maybe worth having something in the migration guide or release notes?
